### PR TITLE
Gazebo updates

### PR DESCRIPTION
**Pull request by JasonGibson274**
Complete overhaul of the IGVC simulation environment.

* The URDF file has been converted to xarco to make it easier to update the file for new robots. 
* Ground truth pose is now correctly published with the current location of the robot being the origin of the odom coordinate frame
* The qualification course has been updated to have increasing difficulty as you move laterally. There is an argument that can be passed from commandline to give different starting locations.

What I need a review for
* Look through the urdf.xacro file and tell me what should be commented more than it already is
* Other general style and commenting how things works

Eventually I will write documentation on how to create a course but it is dumb without proper tooling that I will not get done in the foreseeable future. 

Please ignore any python scripts, they are convenience scripts that will be updated or can be removed if desired.

### CODE REVIEW

File: igvc_gazebo/launch/igvc_control.launch [hunk 1]
```diff
@@ -10,6 +10,13 @@
     args="$(find igvc_gazebo)/config/igvc_joint_control.yaml"/>
 
   <node name="igvc_control" pkg="igvc_gazebo" type="control">
+      <!-- PID values for wheel speed -->
+      <param name="speed_P_left" value="0.5"/>
+      <param name="speed_P_right" value="0.5"/>
+      <param name="speed_D_left" value="3.0"/>
+      <param name="speed_D_right" value="3.0"/>
```
Comment (oswinso) [inline igvc_gazebo/launch/igvc_control.launch]: No I for PID?

File: igvc_gazebo/nodes/ground_truth/main.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    result.twist.twist.linear.x = -msg->twist.twist.linear.y;
+    result.twist.twist.linear.y = msg->twist.twist.linear.x;
+
+    result.header = msg->header;
+    result.child_frame_id = "/base_footprint";
+    result.header.frame_id = "/odom";
+
+    tf::Quaternion quat(msg->pose.pose.orientation.x, msg->pose.pose.orientation.y,
+                        msg->pose.pose.orientation.z, msg->pose.pose.orientation.w);
+   quat *= tf::createQuaternionFromRPY(0, 0, 1.57);
```
Comment (oswinso) [inline igvc_gazebo/nodes/ground_truth/main.cpp]: why not do
```
tf::Quaternion quat;
tf::quaternionMsgToTf(msg->pose.pose.orientation, quat);
quat *= tf::createQuaternionFromYaw(1.57);
tf::quaternionTfToMsg(quat, result.pose.pose.orientation);
```

File: igvc_gazebo/launch/igvc_control.launch (hunk 1 above)
Comment (JasonGibson274) [reply igvc_gazebo/launch/igvc_control.launch]: we didn't used to use it on the actual robot before, so it wasn't implemented.

File: igvc_description/launch/spawn_jaymii.launch [hunk 3]
```diff
@@ -1,17 +1,34 @@
...
+  <arg name="yaw" default="0"/>
+  <!-- accelerate lidar with GPU -->
+  <arg name="gpu" default="false"/>
+  <!-- Spawn the vehicle. -->
+
+  <param name="robot_description" command="$(find xacro)/xacro $(find igvc_description)/urdf/jessii.urdf.xacro gpu:=$(arg gpu)" />
+
+
+  <!-- uncomment below line to change to using the old robot -->
+  <!--<param name="robot_description" command="cat $(find igvc_description)/urdf/jessi.urdf" />-->
```
Comment (oswinso) [inline igvc_description/launch/spawn_jaymii.launch]: Change this from uncomment to an env / arg?

File: igvc_gazebo/nodes/ground_truth/main.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ros::Time last_estimate;
+
+
+void odom_callback(const nav_msgs::Odometry::ConstPtr& msg) {
+  last_estimate = msg->header.stamp;
+}
+
+void ground_truth_callback(const nav_msgs::Odometry::ConstPtr& msg) {
+
+  std::lock_guard<std::mutex> truth_lock(mutex);
```
Comment (oswinso) [inline igvc_gazebo/nodes/ground_truth/main.cpp]: is this mutex necessary when you're only single threaded?

File: igvc_utils/include/igvc_utils/NodeUtils.hpp [hunk 5]
```diff
@@ -14,8 +14,8 @@ void param(const ros::NodeHandle &pNh, const std::string &param_name, T &param_v
 {
   if (!pNh.param(param_name, param_val, default_val))
   {
-    ROS_ERROR_STREAM("Missing parameter " << param_name << " from " << pNh.getNamespace()
-                                          << ". Continuing with default values.");
+    ROS_WARN_STREAM("Missing parameter " << param_name << " from " << pNh.getNamespace()
+                                          << ". Continuing with default values");
```
Comment (oswinso) [inline igvc_utils/include/igvc_utils/NodeUtils.hpp]: Change doesn't belong here I think. Make a separate issue.

File: igvc_utils/launch/rosbag_sensors.launch [hunk 6]
```diff
@@ -1,9 +0,0 @@
-<!-- rosbag.launch -->
-<!--
-    This file will launch the rosbag recorder and will log all topics
-    excluding some of the duplicate image_transport topics.
-    -->
-<launch>
-    <node pkg="rosbag" type="record" name="record" output="screen"
-    args="/battery /encoders /fix /pc2 /scan/pointcloud /imu /joint_states /lights /motors /robot_enabled /tf /tf_static /usb_cam_center/camera_info /usb_cam_center/image_raw/compressed /vel /waypoint /time_reference" />
```
Comment (oswinso) [inline igvc_utils/launch/rosbag_sensors.launch]: Accidental delete or is there something else we can use to rosbag?

Comment (JasonGibson274) [reply igvc_utils/launch/rosbag_sensors.launch]: We don't use this anymore AFIK

File: igvc_gazebo/nodes/ground_truth/main.cpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+#include <tf/transform_broadcaster.h>
+#include <tf/transform_listener.h>
+#include <tf/transform_datatypes.h>
+#include <igvc_utils/NodeUtils.hpp>
+#include <robot_localization/navsat_conversions.h>
+
+ros::Publisher ground_truth_pub;
+std::mutex mutex;
+nav_msgs::Odometry og_pose;
+ros::Time last_estimate;
```
Comment (oswinso) [inline igvc_gazebo/nodes/ground_truth/main.cpp]: globals should almost never be used. When used, they should be prefixed with g_.
http://wiki.ros.org/CppStyleGuide#Global_variables

File: igvc_gazebo/nodes/ground_truth/main.cpp [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+void odom_callback(const nav_msgs::Odometry::ConstPtr& msg) {
```
Comment (oswinso) [inline igvc_gazebo/nodes/ground_truth/main.cpp]: In general, function and class method names are camelCased.

http://wiki.ros.org/CppStyleGuide#Function_.2BAC8_Methods

File: igvc_gazebo/nodes/control/main.cpp [hunk 9]
```diff
@@ -59,13 +55,37 @@ int main(int argc, char **argv)
   ros::init(argc, argv, "jaymii_controller");
 
   ros::NodeHandle handle;
+  ros::NodeHandle pNh("~");
 
   ros::Publisher leftWheelEffortPublisher =
       handle.advertise<std_msgs::Float64>("/left_wheel_effort_controller/command", 1);
   ros::Publisher rightWheelEffortPublisher =
```
Comment (oswinso) [inline igvc_gazebo/nodes/control/main.cpp]: change to `right_wheel_effort_publisher`.

> In general, variable names are under_scored.

http://wiki.ros.org/CppStyleGuide#Variables

File: igvc_description/launch/spawn_jaymii.launch (hunk 3 above)
Comment (JasonGibson274) [reply igvc_description/launch/spawn_jaymii.launch]: I'm going to leave it. I don't think anyone is going to really use the old way the robot would launch.

**Comment by JasonGibson274** [comment]
Alright final review then we can merge. I think everyone is done.

File: igvc_description/urdf/jessii.urdf.xacro [hunk 10]
```diff
@@ -0,0 +1,655 @@
...
+  <joint name="center_cam_optical_joint" type="fixed">
+      <parent link="center_cam"/>
+      <child link="center_cam_optical"/>
+      <origin xyz="0 0 0" rpy="-1.5708 0 -1.5708"/>
+  </joint>
+
+  <gazebo reference="center_cam">
+      <sensor name="center_cam_sensor" type="camera">
+          <camera>
+              <horizontal_fov>1.2290609</horizontal_fov>
```
Comment (Alescontrela) [inline igvc_description/urdf/jessii.urdf.xacro]: are these values correct for the Logitech C920?

File: igvc_gazebo/launch/igvc_control.launch [hunk 11]
```diff
@@ -9,7 +9,17 @@
   <node name="controller_spawner" pkg="controller_manager" type="spawner"
     args="$(find igvc_gazebo)/config/igvc_joint_control.yaml"/>
 
-  <node name="igvc_control" pkg="igvc_gazebo" type="control">
+  <node name="igvc_control" pkg="igvc_gazebo" type="control" output="screen">
+      <!-- PID values for wheel speed -->
+      <param name="speed_P_left" value="0.2"/>
```
Comment (Alescontrela) [inline igvc_gazebo/launch/igvc_control.launch]: These really low values might explain why the robot takes a long time to converge to the path in simulation

File: igvc_description/urdf/jessii.urdf.xacro [hunk 12]
```diff
@@ -0,0 +1,655 @@
...
+  <!-- include other xacro files to use their macros -->
+  <xacro:include filename="$(find velodyne_description)/urdf/VLP-16.urdf.xacro"/>
+
+  <!-- passed in parameters -->
+  <xacro:arg name="gpu" default="false" />
+
+  <!-- constants -->
+  <xacro:property name="M_PI" value="3.1415926535897931" />
+
+  <xacro:property name="chassis_mass" value="20" />
```
[the remark this answers is not in the document]
Comment (JasonGibson274) [reply igvc_description/urdf/jessii.urdf.xacro]: not at all. It drives fine and deoesn't really matter

File: igvc_description/urdf/jessii.urdf.xacro [hunk 13]
```diff
@@ -0,0 +1,655 @@
...
+
+  <!-- wheels -->
+  <xacro:property name="drive_tire_diameter" value="0.305"/>
+  <xacro:property name="drive_tire_width" value="0.0762"/>
+  <!-- TODO set proper mass values -->
+  <xacro:property name="drive_wheel_mass" value="5.0"/>
+
+  <xacro:property name="wheelbase" value="0.714333"/>
+
+  <!-- wheel acuator limits -->
```
[the remark this answers is not in the document]
Comment (JasonGibson274) [reply igvc_description/urdf/jessii.urdf.xacro]: effort and velocity? is that not clear?

File: igvc_description/urdf/jessii.urdf.xacro [hunk 14]
```diff
@@ -0,0 +1,655 @@
...
+  <xacro:property name="drive_wheel_eff_limit" value="8"/>
+  <xacro:property name="drive_wheel_vel_limit" value="20"/>
+
+  <!-- drive shock params-->
```
[the remark this answers is not in the document]
Comment (JasonGibson274) [reply igvc_description/urdf/jessii.urdf.xacro]: We do in sim and were planning on it in real life. I'm not taking them out unless you think it breaks driving.

File: igvc_description/urdf/jessii.urdf.xacro (hunk 10 above)
Comment (JasonGibson274) [reply igvc_description/urdf/jessii.urdf.xacro]: nope

File: igvc_description/urdf/jessii.urdf.xacro [hunk 15]
```diff
@@ -0,0 +1,655 @@
+<?xml version="1.0"?>
+
```
[the remark this answers is not in the document]
Comment (JasonGibson274) [reply igvc_description/urdf/jessii.urdf.xacro]: I am not sure what you mean.

[the remark this answers is not in the document]
Comment (Alescontrela) [reply igvc_description/urdf/jessii.urdf.xacro]: Typo. What I mean is that there's a bunch of transforms defined in this file, i.e. 

```C++
xacro:property name="right_camera_y" value="0.3"/>
  <xacro:property name="right_camera_x" value="0.216371"/>
  <xacro:property name="right_camera_pitch" value="0.0"/>
  <xacro:property name="right_camera_yaw" value="0.0"/>
  <xacro:property name="right_camera_z" value="1.354330"/>
```

but it's not clear what they're defined relative to...

File: igvc_description/urdf/jessii.urdf.xacro (hunk 10 above)
Comment (JasonGibson274) [reply igvc_description/urdf/jessii.urdf.xacro]: This was actually right

File: igvc_gazebo/launch/igvc_control.launch (hunk 11 above)
Comment (JasonGibson274) [reply igvc_gazebo/launch/igvc_control.launch]: It ramps up equivalent to what I saw with Oswin's step inputs. You can change the PID values though. Separate PR IMO.

File: igvc_description/urdf/jessii.urdf.xacro (hunk 10 above)
Comment (JasonGibson274) [reply igvc_description/urdf/jessii.urdf.xacro]: It was in radians. I added a comment

**Comment by Alescontrela** [comment]
LGTM :+1:

**Pull request merged**